File: api/tests/announcement_test.py
""" Test for announcement APIs """
import requests

from tests import SERVER, ADMIN_ACCESS, BASE_ACCESS
from tests.framework import assert_status_code

def add_announcement():
    """ Add a new announcement """
    response = requests.get(SERVER+"/api/department/1/announcements",
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={})
    assert_status_code(response, 200)
    data = response.json()
    initial_ids = []
    for item in data['data']:
        initial_ids.append(item['id'])

    response = requests.post(SERVER+"/api/announcement/1",
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={'Scope':'2',
                         'Text': 'testing',
                         'Email': '0'})
    assert_status_code(response, 201)

    response = requests.get(SERVER+"/api/department/1/announcements",
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={})
    assert_status_code(response, 200)
    data = response.json()
    # Find New Index
    target = None
    for item in data['data']:
        if target is None and item['id'] not in initial_ids:
            target = item['id']

    response = requests.get(SERVER+"/api/announcement/"+target,
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={})
    assert_status_code(response, 200)
    item = response.json()
    assert 'type' in item
    assert item['type'] == 'announcement'
    assert 'id' in item
    assert item['id'] == target
    assert 'departmentId' in item
    assert item['departmentId'] == '1'
    assert 'postedOn' in item
    assert 'postedBy' in item
    assert 'scope' in item
    assert item['scope'] == '2'
    assert 'text' in item
    assert item['text'] == 'testing'

    return target

def test_announce_priv():
    """ Getting the announcments as a privilaged user """

    target = add_announcement()

    response = requests.get(SERVER+"/api/department/1/announcements",
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={})
    assert_status_code(response, 200)
    data = response.json()
    assert 'data' in data
    assert 'type' in data
    assert data['type'] == 'announcement_list'
    item = data['data'][0]
    assert 'type' in item
    assert item['type'] == 'announcement'
    assert 'id' in item
    assert 'departmentId' in item
    assert item['departmentId'] == '1'
    assert 'postedOn' in item
    assert 'postedBy' in item
    assert 'scope' in item
    assert 'text' in item

    response = requests.put(SERVER+"/api/announcement/"+target,
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={'Text': 'New Message'})
    assert_status_code(response, 200)

    response = requests.get(SERVER+"/api/announcement/"+target,
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={})
    assert_status_code(response, 200)
    item = response.json()
    assert 'type' in item
    assert item['type'] == 'announcement'
    assert 'id' in item
    assert item['id'] == target
    assert 'departmentId' in item
    assert item['departmentId'] == '1'
    assert 'postedOn' in item
    assert 'postedBy' in item
    assert 'scope' in item
    assert item['scope'] == '2'
    assert 'text' in item
    assert item['text'] == 'New Message'

    response = requests.delete(SERVER+"/api/announcement/"+target,
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={})
    assert_status_code(response, 204)
    assert response.text == ''


def test_announce_unpriv():
    """ Getting the announcments as a unprivilaged user """
    response = requests.get(SERVER+"/api/department/1/announcements",
                   headers={'Authorization': 'Bearer '+BASE_ACCESS},
                   data={})
    assert_status_code(response, 200)
    data = response.json()
    assert 'data' in data
    assert 'type' in data
    assert data['type'] == 'announcement_list'
    if len(data['data']) > 0:
        item = data['data'][0]
        assert 'type' in item
        assert item['type'] == 'announcement'
        assert 'id' in item
        assert 'departmentId' in item
        assert item['departmentId'] == '1'
        assert 'postedOn' in item
        assert 'postedBy' in item
        assert 'scope' in item
        assert 'text' in item
    initial_ids = []
    for item in data['data']:
        initial_ids.append(item['id'])

    if len(initial_ids) > 0 :
        response = requests.post(SERVER+"/api/announcement/1",
                       headers={'Authorization': 'Bearer '+BASE_ACCESS},
                       data={'Scope':'2',
                             'Text': 'testing',
                             'Email': '0'})
        assert_status_code(response, 403)

        response = requests.put(SERVER+"/api/announcement/"
                                +initial_ids[0],
                       headers={'Authorization': 'Bearer '+BASE_ACCESS},
                       data={'Text': 'New Message'})
        assert_status_code(response, 403)

        response = requests.delete(SERVER+"/api/announcement/"
                                   +initial_ids[0],
                       headers={'Authorization': 'Bearer '+BASE_ACCESS},
                       data={})
        assert_status_code(response, 403)

def test_announce_not_found():
    """ Getting the announcments with invalid id"""
    response = requests.get(SERVER+"/api/department/1/announcements",
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={})
    assert_status_code(response, 200)
    data = response.json()
    initial_ids = []
    for item in data['data']:
        initial_ids.append(item['id'])

    target = None
    for i in range(0, len(initial_ids) + 1):
        if target is None and i not in initial_ids:
            target = str(i)

    response = requests.post(SERVER+"/api/announcement",
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={'Scope':'2',
                         'Text': 'testing',
                         'Email': '0'})
    assert_status_code(response, 404)

    response = requests.get(SERVER+"/api/announcement/"+target,
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={})
    assert_status_code(response, 404)

    response = requests.put(SERVER+"/api/announcement/"+target,
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={'Text': 'New Message'})
    assert_status_code(response, 404)

    response = requests.delete(SERVER+"/api/announcement/"+target,
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={})
    assert_status_code(response, 404)

def test_announce_params():
    """ Getting the announcments with invalid params"""

    response = requests.post(SERVER+"/api/announcement/1",
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={'Text': 'testing',
                         'Email': '0'})
    assert_status_code(response, 400)

    response = requests.post(SERVER+"/api/announcement/1",
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={'Scope': '2',
                         'Email': '0'})
    assert_status_code(response, 400)

    response = requests.post(SERVER+"/api/announcement/1",
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={'Email': '0'})
    assert_status_code(response, 400)

    target = add_announcement()

    response = requests.put(SERVER+"/api/announcement/"+target,
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={})
    assert_status_code(response, 400)

    response = requests.put(SERVER+"/api/announcement/"+target,
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={'Department': '-1'})
    assert_status_code(response, 400)

    response = requests.delete(SERVER+"/api/announcement/"+target,
                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
                   data={})
    assert_status_code(response, 204)

File: api/src/Tests/TestCase/Controller/AnnouncementTest.php
<?php

namespace App\Tests\TestCase\Controller\Announcement;

use App\Controller\Announcement\GetAnnouncement;
use PHPUnit\Framework\TestCase;
use App\Tests\Traits\AppTestTrait;
use Slim\Http\Response;

class AnnouncementTest extends TestCase
{

    use AppTestTrait;


    protected function addAnnouncement($scope): string
    {
        $request = $this->createRequest('GET', '/department/1/announcements');
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 200);

        $data = json_decode((string)$response->getBody());
        $initial_ids = [];
        foreach ($data->data as $item) {
            $initial_ids[] = $item->id;
        }

        $request = $this->createRequest('POST', '/announcement/1');
        $request = $request->withParsedBody(['Scope' => $scope,
                                             'Text' => 'testing',
                                             'Email' => 1]);
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 201);

        $request = $this->createRequest('GET', '/department/1/announcements');
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 200);

        $data = json_decode((string)$response->getBody());
        # Find New Index
        $target = null;
        foreach ($data->data as $item) {
            if ($target == null && !in_array($item->id, $initial_ids)) {
                $target = $item->id;
                $this->assertSame([
                    'type' => 'announcement',
                    'id' => $target,
                    'departmentId' => '1',
                    'postedOn' => $item->postedOn,
                    'postedBy' => '1000',
                    'scope' => "$scope",
                    'text' => 'testing'
                ], (array)$item);
            }
        }

        $request = $this->createRequest(
            'GET',
            '/announcement/'.$target,
            "fields=type,id,departmentId,postedBy,scope,text"
        );
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 200);
        $data = json_decode((string)$response->getBody());
        $this->assertSame([
            'type' => 'announcement',
            'id' => $target,
            'departmentId' => '1',
            'postedBy' => '1000',
            'scope' => "$scope",
            'text' => 'testing'
        ], (array)$data);

        return $target;

    }


    public function testAnnounce(): void
    {
        $target = $this->addAnnouncement(1);

        $request = $this->createRequest(
            'GET',
            '/department/1/announcements',
            'include=departmentId,postedBy'
        );
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 200);
        $data = json_decode((string)$response->getBody());
        $this->assertNotEmpty($data->data);

        $request = $this->createRequest('PUT', '/announcement/'.$target);
        $request = $request->withParsedBody([
            'Department' => 2,
            'Text' => 'New Message',
            'Scope' => 1
        ]);
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 200);

        $request = $this->createRequest(
            'GET',
            '/announcement/'.$target,
            'include=departmentId,postedBy'
        );
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 200);
        $data = json_decode((string)$response->getBody());
        $this->assertSame($data->text, 'New Message');

        $request = $this->createRequest(
            'GET',
            '/member/1000/announcements',
            'include=departmentId,postedBy'
        );
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 200);
        $data = json_decode((string)$response->getBody());
        $this->assertNotEmpty($data->data);

        $request = $this->createRequest('DELETE', '/announcement/'.$target);
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 204);

    }


    public function testAnnounceErrors(): void
    {
        $target = $this->addAnnouncement(2);

        $request = $this->createRequest(
            'GET',
            '/department/-1/announcements'
        );
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 404);

        $request = $this->createRequest('PUT', '/announcement/'.$target);
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 400);

        $request = $this->createRequest('PUT', '/announcement/-1');
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 404);

        $request = $this->createRequest('PUT', '/announcement/'.$target);
        $request = $request->withParsedBody([
            'Department' => -1,
        ]);
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 400);

        $request = $this->createRequest('POST', '/announcement/1');
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 400);

        $request = $this->createRequest('POST', '/announcement/-1');
        $request = $request->withParsedBody(['Scope' => 2,
                                             'Text' => 'testing',
                                             'Email' => 0]);
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 404);

        $request = $this->createRequest('POST', '/announcement/1');
        $request = $request->withParsedBody(['Text' => 'testing',
                                             'Email' => 0]);
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 400);

        $request = $this->createRequest('POST', '/announcement/1');
        $request = $request->withParsedBody(['Scope' => 2,
                                             'Email' => 0]);
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 400);

        $request = $this->createRequest('DELETE', '/announcement/'.$target);
        $this->container['request'] = $request;
        $response = $this->app->run(true);
        $this->assertSame($response->getStatusCode(), 204);

    }


    /* End */
}
